### ISSUE TITLE
[META] Improve test suite

### ISSUE THREAD

**Issue by juanignaciosl** [CartoDB/cartodb#6754]
- [ ] Improve factories for core (visualizations, maps, etc). Now [some helpers even use the endpoints](https://github.com/CartoDB/cartodb/blob/master/spec/factories/visualization_creation_helpers.rb), which is an overkill and hard to reuse.
- [x] Improve speed.
- [ ] Automate smoke tests.
- [ ] Cover layer
  - Controller: #6759
  - Model.

**Comment by javitonino** [comment]
A list of suspect tests, using the branch `test_profiling`. They are run as:
`PROFILE=1 bundle exec rspec -p --require ./spec/rspec_configuration.rb $spec`

Failing tests
=============
Note that some of this fail because the profiler adds some dummy rows to each table in the DB, and some of the tests do things like `Tag.count` to get a global object count (which they probably shouldn't do).
 - [ ] spec/models/data_import_spec.rb (22)
 - [ ] services/table-geocoder/spec/table_geocoder_spec.rb (15)
 - [ ] spec/requests/signup_controller_spec.rb (14)
 - [ ] spec/requests/admin/pages_controller_spec.rb (13)
 - [ ] spec/models/permission_spec.rb (12)
 - [ ] spec/models/user_organization_spec.rb (7)
 - [ ] spec/requests/admin/organization_users_controller_spec.rb (6)
 - [ ] services/importer/spec/acceptance/rar_spec.rb (5)
 - [ ] spec/models/visualization/organization_visualization_spec.rb (4)
 - [ ] spec/requests/sessions_controller_spec.rb (4)
 - [ ] services/table-geocoder/spec/lib/abstract_table_geocoder_spec.rb (4)
 - [ ] services/importer/spec/acceptance/gpx_spec.rb (3)
 - [ ] services/importer/spec/acceptance/kml_spec.rb (3)
 - [ ] spec/requests/carto/api/visualizations_controller_spec.rb (3)
 - [ ] spec/requests/admin/users_controller_spec.rb (3)
 - [ ] spec/models/table_spec.rb (2)
 - [ ] spec/requests/sessions_spec.rb (2)
 - [ ] services/importer/spec/acceptance/csv_spec.rb (2)
 - [ ] services/importer/spec/unit/shp_normalizer_spec.rb (2)
 - [ ] services/importer/spec/unit/kml_splitter_spec.rb (2)
 - [ ] services/table-geocoder/spec/geocoder_cache_spec.rb (2)
 - [ ] spec/models/user_presenter_spec.rb (1)
 - [ ] spec/requests/carto/api/user_creations_controller_spec.rb (1)
 - [ ] spec/lib/explore_api_spec.rb (1)
 - [ ] services/importer/spec/unit/csv_normalizer_spec.rb (1)
 - [ ] services/importer/spec/unit/unp_spec.rb (1)
 - [ ] spec/queries/carto/visualization_query_builder_spec.rb (1)
 - [ ] spec/requests/carto/api/organization_users_controller_spec.rb (1)
 - [ ] spec/requests/api/imports_spec.rb (1)
 - [ ] services/table-geocoder/spec/lib/gme/table_geocoder_spec.rb (1)
 - [ ] spec/models/synchronization/collection_spec.rb (1)
 - [ ] spec/models/synchronization/synchronization_oauth_spec.rb (1)
 - [ ] spec/models/shared_entity_spec.rb (1)
 - [ ] spec/requests/superadmin/users_spec.rb (1)
 - [ ] spec/requests/superadmin/organizations_spec.rb (1)
 - [ ] spec/models/carto/user_creation_spec.rb (1)


Tests that delete more objects that they create
===============================================
e.g: `Carto::Widget.each(&:destroy)`

 - [ ] spec/models/user_presenter_spec.rb (Carto::Geocoding)
 - [ ] spec/requests/carto/api/templates_controller_spec.rb (Carto::Template)
 - [ ] spec/requests/carto/api/widgets_controller_spec.rb (Carto::Widget)
 - [ ] spec/models/carto/template_spec.rb (Carto::Template)
 - [ ] spec/services/carto/visualizations_export_service_spec.rb (Carto::VisualizationBackup)
 - [ ] spec/queries/carto/visualization_query_builder_spec.rb (Carto::ExternalDataImport)
 - [ ] spec/requests/carto/api/connectors_controller_spec.rb (Carto::FeatureFlag, Carto::FeatureFlagsUser)
 - [ ] spec/requests/carto/api/analyses_controller_spec.rb (Carto::FeatureFlag, Carto::FeatureFlagsUser)
 - [ ] spec/services/carto/user_table_index_service_spec.rb (Carto::Widget)
 - [ ] spec/requests/carto/api/mapcaps_controller_spec.rb (Carto::FeatureFlag, Carto::FeatureFlagsUser)


Pending specs
=============
 - [ ] spec/requests/api/visualizations_spec.rb (16)
 - [ ] spec/models/user_spec.rb (4)
 - [ ] services/geocoder/spec/geocoder_spec.rb (4)
 - [ ] spec/models/layer_spec.rb (3)
 - [ ] spec/requests/carto/api/connectors_controller_spec.rb (3)
 - [ ] spec/models/table_spec.rb (2)
 - [ ] services/twitter-search/spec/unit/json_to_csv_converter_spec.rb (2)
 - [ ] spec/requests/superadmin/organizations_spec.rb (2)
 - [ ] spec/models/geocoding_spec.rb (1)
 - [ ] spec/lib/api_calls_spec.rb (1)
 - [ ] spec/lib/cartodb/connection_pool_spec.rb (1)
 - [ ] services/importer/spec/acceptance/osm_spec.rb (1)
 - [ ] services/importer/spec/acceptance/sql_spec.rb (1)
 - [ ] services/importer/spec/acceptance/raster2pgsql_spec.rb (1)
 - [ ] services/importer/spec/unit/shp_helper_spec.rb (1)
 - [ ] spec/requests/api/imports_spec.rb (1)
 - [ ] spec/requests/api/assets_spec.rb (1)
 - [ ] services/table-geocoder/spec/lib/gme/table_geocoder_spec.rb (1)
 - [ ] services/datasources/spec/acceptance/dropbox_spec.rb (1)
 - [ ] services/datasources/spec/acceptance/gdrive_spec.rb (1)
 - [ ] services/datasources/spec/integration/twitter_spec.rb (1)
 - [ ] spec/requests/superadmin/users_spec.rb (1)
 - [ ] spec/requests/carto/api/invitations_controller_spec.rb (1)

Tests that create a lot of objects
==================================
More than 10 users:
 - [ ] spec/models/user_spec.rb (117)
 - [ ] spec/models/carto/user_spec.rb (56)
 - [ ] spec/requests/carto/api/visualizations_controller_spec.rb (49)
 - [ ] spec/models/map_spec.rb (35)
 - [ ] spec/requests/superadmin/users_spec.rb (27)
 - [ ] spec/models/layer_spec.rb (26)
 - [ ] spec/models/organization_spec.rb (24)
 - [ ] spec/models/carto/user_creation_spec.rb (24)
 - [ ] spec/requests/superadmin/organizations_spec.rb (18)
 - [ ] spec/models/user_organization_spec.rb (18)
 - [ ] spec/requests/carto/builder/visualizations_controller_spec.rb (17)
 - [ ] spec/services/carto/visualizations_export_service_2_spec.rb (15)
 - [ ] spec/requests/admin/organization_users_controller_spec.rb (13)

More than 10 organizations:
 - [ ] spec/models/user_spec.rb (45)
 - [ ] spec/models/carto/user_spec.rb (22)
 - [ ] spec/models/user_organization_spec.rb (18)
 - [ ] spec/requests/carto/api/visualizations_controller_spec.rb (13)
 - [ ] spec/requests/superadmin/organizations_spec.rb (13)
 - [ ] spec/models/organization_spec.rb (12)

More than 30 visualizations:
 - [ ] spec/requests/superadmin/users_spec.rb (600)
 - [ ] spec/models/table_spec.rb (184)
 - [ ] spec/requests/admin/pages_controller_spec.rb (150)
 - [ ] spec/services/carto/visualizations_export_service_2_spec.rb (52)
 - [ ] spec/models/visualization/collection_spec.rb (41)
 - [ ] spec/models/map_spec.rb (40)
 - [ ] spec/connectors/importer_spec.rb (35)
 - [ ] spec/requests/admin/visualizations_spec.rb (33)
 - [ ] spec/requests/carto/api/visualizations_controller_spec.rb (30)

More than 30 user tables
 - [ ] spec/models/table_spec.rb (180)
 - [ ] spec/requests/carto/api/visualizations_controller_spec.rb (48)
 - [ ] spec/models/map_spec.rb (39)


Slow test cases (> 10 s)
========================
 - [ ] 11.46 spec/requests/carto/builder/visualizations_controller_spec.rb
 - [ ] 11.2 spec/lib/cartodb/stats/platform_spec.rb
 - [ ] 20.68 spec/lib/cartodb/connection_pool_spec.rb
 - [ ] 18.85 spec/lib/cartodb/connection_pool_spec.rb
 - [ ] 42.59 services/importer/spec/acceptance/shp_spec.rb
 - [ ] 11.53 services/importer/spec/unit/downloader_spec.rb
 - [ ] 11.53 services/importer/spec/unit/runner_spec.rb
 - [ ] 29.83 spec/connectors/importer_spec.rb
 - [ ] 29.72 spec/connectors/importer_spec.rb
 - [ ] 10.07 spec/connectors/importer_overviews_spec.rb
 - [ ] 17.71 spec/requests/carto/api/layers_controller_spec.rb
 - [ ] 12.86 spec/models/synchronization/member_spec.rb
 - [ ] 19.26 spec/models/organization_spec.rb
 - [ ] 15.61 spec/models/organization_spec.rb
 - [ ] 12.22 spec/models/organization_spec.rb
 - [ ] 11.47 spec/models/carto/user_migration_spec.rb

Slow specs
==========
 - [ ] spec/models/user_spec.rb (9 m)
 - [ ] spec/requests/carto/api/visualizations_controller_spec.rb (7 m)
 - [ ] spec/models/carto/user_spec.rb (4 m)
 - [ ] spec/models/map_spec.rb (3.5 m)
 - [ ] spec/requests/superadmin/users_spec.rb (3 m)
 - [ ] spec/models/layer_spec.rb (2.5 m)
 - [ ] spec/requests/api/imports_spec.rb (2.5 m)
 - [ ] spec/connectors/importer_spec.rb (2.5 m)
 - [ ] spec/models/table_spec.rb (2.5 m)
 - [ ] spec/models/organization_spec.rb (2 m)
 - [ ] spec/models/user_organization_spec.rb (2 m)
 - [ ] services/user-mover/spec/user_mover_spec.rb (2 m)
 - [ ] spec/requests/carto/builder/visualizations_controller_spec.rb (2 m)

**Comment by xavijam** [comment]
There has been no activity on this issue for more several months. We are closing it. If you think this still needs to be addressed please open a new issue.

**Comment by javitonino** [comment]
Reviving. The last comment came from a recent leapfrog and I intend to fix the listed tests in the following weeks.

**Comment by nobuti** [comment]
There has been no activity on this issue for more several months. We are closing it. If you think this still needs to be addressed please open a new issue.

**Comment by nobuti** [comment]
There has been no activity on this issue for more several months. We are closing it. If you think this still needs to be addressed please open a new issue.

**Comment by javitonino** [comment]
Damm @nobuti bot, stop closing my issues! :laughing:

**Comment by rafatower** [comment]
In the past, failures of tests run in parallel were mostly due to test ordering, interfering with one another (you know in ruby you can mock everything and sometimes the mock remains between test executions).

I'd say it would be helpful to either force a giving ordering within each environment, or to log it (and maybe keep it as an "artifact" attached to the build in jenkins).

**Comment by javitonino** [comment]
I am going to close this one since https://github.com/CartoDB/cartodb/pull/11636 addresses a lot of the issues in this ticket. I'll start a fresh list of issues after that PR is deployed.
